staging/kos/pkg/controllers/subnet/validator.go

```diff
@@ -0,0 +1,519 @@
+	// s.
+	oldRivals := v.updateConflictsCache(sd)
+	for _, r := range oldRivals {
+		v.queue.Add(r)
+	}
+
+	// When a subnet is validated, in case of conflicts it is added to the
+	// conflicts caches of its rival subnets, so that when such rivals are
+	// deleted it can be revalidated. This is useless for a subnet that has
+	// already passed validation. Also, if this controllers sees
+	// s.Status.Validated to be true, there's no way that such value can refer
+	// to a previous, stale version of the subnet, because the API server clears
+	// s.Status.Validated when a subnet spec is updated in a way that requires
+	// revalidation. Hence, we can simply stop processing here if
+	// s.Status.Validated is true.
+	// ? Maybe we should make it clearer somewhere (e.g. API types comments and
+	// doc) that the API server clears s.Status.Validated when a subnet spec is
+	// updated in a way that requires revalidation.
+	if s.Status.Validated {
+		return nil
+	}
+
+	// Retrieve all the other subnets. Doing a live list as opposed to a
+	// cache-based one (through the informer) prevents race conditions that can
+	// arise in case of multiple validators running.
+	allSubnets, err := v.netIfc.Subnets(k8scorev1api.NamespaceAll).List(k8smetav1.ListOptions{})
+	if err != nil && doNotRetryList(err) {
+		glog.Errorf("live list of all subnets against API server failed while validating %s: %s. There will be no retry because of the nature of the error", sd.namespacedName, err.Error())
+		// This should never happen, no point in retrying.
+		return nil
+	}
+	if err != nil {
+		return fmt.Errorf("live list of all subnets against API server failed: %s", err.Error())
+	}
+
+	// Look for conflicts with all the other subnets and record them in the
+	// rivals conflicts caches.
+	conflictsMsgs, conflictFound, err := v.recordConflicts(sd, allSubnets.Items)
+	if err != nil {
+		return err
+	}
+
+	if err := v.updateSubnetValidity(s, !conflictFound, conflictsMsgs); err != nil {
+		return fmt.Errorf("failed to write validation outcome into %s's status: %s", sd.namespacedName, err.Error())
+	}
+
+	return nil
+}
+
+func (v *Validator) clearStaleRV(s k8stypes.NamespacedName) {
+	v.staleRVsMutex.Lock()
+	defer v.staleRVsMutex.Unlock()
+
+	delete(v.staleRVs, s)
+}
+
+func (v *Validator) clearConflictsCache(s k8stypes.NamespacedName) []k8stypes.NamespacedName {
+	v.conflictsMutex.Lock()
+	defer v.conflictsMutex.Unlock()
+
+	if c := v.conflicts[s]; c != nil {
+		delete(v.conflicts, s)
+		return c.rivals
+	}
+
+	return nil
+}
+
+func parseSubnet(s *netv1a1.Subnet) (*subnetData, error) {
+	_, ipNet, err := net.ParseCIDR(s.Spec.IPv4)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse %#+v: %s", s, err.Error())
+	}
+	sd := &subnetData{
+		namespacedName: k8stypes.NamespacedName{
+			Namespace: s.Namespace,
+			Name:      s.Name,
+		},
+		vni:   s.Spec.VNI,
+		baseU: ipv4ToUint32(ipNet.IP),
+	}
+	ones, bits := ipNet.Mask.Size()
+	delta := uint32(uint64(1)<<uint(bits-ones) - 1)
```

Comment:
why not put this before the constructor, and set `lastU` in the constructor?

Reply:
Done in 961e55d.